sklearn/metrics/cluster/tests/test_supervised.py

```diff
@@ -456,8 +456,8 @@ def test_adjusted_rand_score_overflow():
     https://github.com/scikit-learn/scikit-learn/issues/20305
     """
     rng = np.random.RandomState(0)
-    y_true = rng.randint(0, 2, 10_000_000, dtype=np.int8)
```

Comment:
I don't know if this would actually trigger the original bug.

Reply:
If we have many workers, using too much memory in a test can cause the OS to kill the pytest-xdist worker. But This is only (10MB) so that seems strange.

Reply:
I ran

```bash
pytest sklearn/metrics/cluster/tests/test_supervised.py::test_adjusted_rand_score_overflow
```

on a raspberry pi b+ with 1gb ram and it crashed.

I think is the `adjusted_rand_score` computation itself that is using a lot of memory. With memory profiling on I saw that:

https://github.com/scikit-learn/scikit-learn/blob/d5e045fd600bf7f3edd984fe7f8599124a67f7f6/sklearn/metrics/cluster/_supervised.py#L143-L150

was using a lot of memory. There is a bunch of copying going on here with the construction of the coo matrix and then a conversion to csr.

Reply:
Running `python -m memory_profiler test.py` on:

```python
from sklearn.metrics.cluster import contingency_matrix
import numpy as np

labels_true = np.random.randint(0, 2, 10_000_000, dtype=np.int8)
labels_pred = np.random.randint(0, 2, 10_000_000, dtype=np.int8)

adjusted_rand_score(labels_true, labels_pred)
```

and profiling `contingency_matrix` I get:

```
   134  142.891 MiB    0.000 MiB           1       if eps is not None and sparse:
   135                                                 raise ValueError("Cannot set 'eps' when sparse=True")
   136                                         
   137  381.641 MiB  238.750 MiB           1       classes, class_idx = np.unique(labels_true, return_inverse=True)
   138  457.938 MiB   76.297 MiB           1       clusters, cluster_idx = np.unique(labels_pred, return_inverse=True)
   139  457.938 MiB    0.000 MiB           1       n_classes = classes.shape[0]
   140  457.938 MiB    0.000 MiB           1       n_clusters = clusters.shape[0]
   141                                             # Using coo_matrix to accelerate simple histogram calculation,
   142                                             # i.e. bins are consecutive integers
   143                                             # Currently, coo_matrix is faster than histogram2d for simple cases
   144  572.469 MiB  114.531 MiB           2       contingency = sp.coo_matrix(
   145  457.938 MiB    0.000 MiB           1           (np.ones(class_idx.shape[0]), (class_idx, cluster_idx)),
   146  457.938 MiB    0.000 MiB           1           shape=(n_classes, n_clusters),
   147  457.938 MiB    0.000 MiB           1           dtype=dtype,
   148                                             )
   149  572.469 MiB    0.000 MiB           1       if sparse:
   150  801.469 MiB  229.000 MiB           1           contingency = contingency.tocsr()
   151  801.469 MiB    0.000 MiB           1           contingency.sum_duplicates()
   152                                             else:
   153                                                 contingency = contingency.toarray()
   154                                                 if eps is not None:
   155                                                     # don't use += as contingency is integer
   156                                                     contingency = contingency + eps
   157  801.469 MiB    0.000 MiB           1       return contingency
```

Reply:
Thanks for checking!